### PR TITLE
[ZEPPELIN-2761] - Neo4j Interpreter

### PR DESCRIPTION
**Pull request by conker84**
### What is this PR for?
This contribution would to introduce Neo4j Cypher intepreter and at the same time provides base APIs that enable other graph databases (or graph framworks such as GraphX or Giraph).

### What type of PR is it?
[Feature]

### Todos
* [x] - Provide base APIs to manage graph results (under: org.apache.zeppelin.interpreter.graph.*)
* [x] - Create the Neo4j intepreter

### What is the Jira issue?
[[ZEPPELIN-2761]](https://issues.apache.org/jira/browse/ZEPPELIN-2761)

### How should this be tested?
Donwload and execute [Neo4j](https://neo4j.com/download/?ref=home) v3.x, you can also pull a [Docker image](https://neo4j.com/developer/docker/).

In order to execute test cases, if you are running Java 7, you need to also provide an environment variable telling the tests where to find Java 8, because Neo4j-the-database needs it to run.
```bash
export NEO4J_JAVA=<path/to/java/home>
```

Use this statement to create a dummy dataset
```bash
%neo4j
UNWIND range(1,100) as id
CREATE (p:Person {id:id, name: "Name " + id, age: id % 3}) WITH collect(p) as people
UNWIND people as p1
UNWIND range(1,10) as friend
WITH p1, people[(p1.id + friend) % size(people)] as p2
CREATE (p1)-[:KNOWS {years: abs(p2.id - p2.id)}]->(p2)
```

Then you can write some simple queries like:
```bash
%neo4j
MATCH (p:Person)-[r:KNOWS]-(p1:Person)
RETURN p, r, p1
LIMIT 10;
```

```bash
%neo4j
MATCH (p:Person)-[r:KNOWS]-(p1:Person)
RETURN p.id AS ID_A, p.name AS NAME_A, r.years AS YEARS, p1.id AS ID_B, p1.name AS NAME_B
LIMIT 20;
```

### Video
zeppelin query examples

### Questions:
* Does the licenses files need update? **Yes**
* Is there breaking changes for older versions? **No**
* Does this needs documentation? **Yes**

### CODE REVIEW

**Comment by jexp** [comment]
I looked through the code and the UX and can confirm that this is great work from a Neo4j point of view.

**Comment by 1ambda** [comment]
Thanks for the great feature! since this is a big change, I guess it takes a time to review and requires multiple reviewers. Let me take a look.

File: docs/index.md [hunk 1]
```diff
@@ -146,6 +146,7 @@ limitations under the License.
   * [Lens](./interpreter/lens.html)
   * [Livy](./interpreter/livy.html)
   * [markdown](./interpreter/markdown.html)
+  * [Neo4j](./interpreter/neo4j.html)
```
Comment (1ambda) [inline docs/index.md]: We need to add the index into https://github.com/apache/zeppelin/blob/master/docs/_includes/themes/zeppelin/_navigation.html#L138 as well

Otherwise, it will not appear in navbar

File: zeppelin-web/karma.conf.js [hunk 2]
```diff
@@ -59,7 +59,6 @@ module.exports = function(config) {
       'bower_components/ace-builds/src-noconflict/mode-python.js',
       'bower_components/ace-builds/src-noconflict/mode-sql.js',
       'bower_components/ace-builds/src-noconflict/mode-markdown.js',
-      'bower_components/ace-builds/src-noconflict/mode-pig.js',
```
Comment (1ambda) [inline zeppelin-web/karma.conf.js]: Any reason to remove `mode-pig.js`?

File: docs/index.md (hunk 1 above)
Comment (conker84) [inline docs/index.md]: Done!

**Comment by 1ambda** [comment]
This is a new feature that makes utilizing graph type easily. Any other opinions?

**Comment by jexp** [comment]
@1ambda any update on the review progress? 
We really would love to see this feature in Apache Zeppelin :)

Thanks a lot.

File: neo4j/src/main/java/org/apache/zeppelin/graph/neo4j/Neo4jConnectionManager.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    this.config = Config.build()
+          .withMaxIdleSessions(Integer.parseInt(properties.getProperty(NEO4J_MAX_CONCURRENCY)))
+          .toConfig();
+    String authType = properties.getProperty(NEO4J_AUTH_TYPE);
+    switch (Neo4jAuthType.valueOf(authType.toUpperCase())) {
+      case BASIC:
+        String username = properties.getProperty(NEO4J_AUTH_USER);
+        String password = properties.getProperty(NEO4J_AUTH_PASSWORD);
+        LOGGER.debug("Creating a BASIC authentication to neo4j with user '{}' and password '{}'",
+                username, password);
```
Comment (felixcheung) [inline neo4j/src/main/java/org/apache/zeppelin/graph/neo4j/Neo4jConnectionManager.java]: probably shouldn't log password?

File: zeppelin-web/src/index.html [hunk 4]
```diff
@@ -140,7 +140,6 @@
     <script src="bower_components/ace-builds/src-noconflict/mode-python.js"></script>
     <script src="bower_components/ace-builds/src-noconflict/mode-sql.js"></script>
     <script src="bower_components/ace-builds/src-noconflict/mode-markdown.js"></script>
-    <script src="bower_components/ace-builds/src-noconflict/mode-pig.js"></script>
```
Comment (felixcheung) [inline zeppelin-web/src/index.html]: why is this removed?

File: neo4j/src/main/resources/interpreter-setting.json [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      "neo4j.auth.user": {
+        "envName": null,
+        "propertyName": "neo4j.auth.user",
+        "defaultValue": "neo4j",
+        "description": "The Neo4j user name."
+      },
+      "neo4j.auth.password": {
+        "envName": null,
+        "propertyName": "neo4j.auth.password",
+        "defaultValue": "neo4j",
```
Comment (felixcheung) [inline neo4j/src/main/resources/interpreter-setting.json]: hmm, would it be better if the default value for password is `""`?

File: neo4j/src/main/java/org/apache/zeppelin/graph/neo4j/Neo4jCypherInterpreter.java [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        for (Entry<String, Object> entry : map.entrySet()) {
+          setTabularResult(String.format(MAP_KEY_TEMPLATE, key, entry.getKey()), entry.getValue(),
+                columns, line, typeSystem);
+        }
+      } else {
+        addValueToLine(key, columns, line, value);
+      }
+    } else if (obj instanceof Map) {
+      @SuppressWarnings("unchecked")
+      Map<String, Object> map = (Map<String, Object>) obj;
```
Comment (felixcheung) [inline neo4j/src/main/java/org/apache/zeppelin/graph/neo4j/Neo4jCypherInterpreter.java]: why do we need unchecked here?

File: neo4j/src/main/java/org/apache/zeppelin/graph/neo4j/Neo4jCypherInterpreter.java [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+      }
+    }
+    return types;
+  }
+
+  @Override
+  public InterpreterResult interpret(String cypherQuery, InterpreterContext interpreterContext) {
+    logger.info("Opening session");
+    if (StringUtils.isEmpty(cypherQuery)) {
+      return new InterpreterResult(Code.ERROR, "Cypher query is Empty");
```
Comment (felixcheung) [inline neo4j/src/main/java/org/apache/zeppelin/graph/neo4j/Neo4jCypherInterpreter.java]: could you check other interpeters? it might be preferred to do nothing if the query is empty.

File: neo4j/src/main/java/org/apache/zeppelin/graph/neo4j/Neo4jCypherInterpreter.java [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+        types.add(record.get("relationshipType").asString());
+      }
+    }
+    return types;
+  }
+
+  @Override
+  public InterpreterResult interpret(String cypherQuery, InterpreterContext interpreterContext) {
+    logger.info("Opening session");
+    if (StringUtils.isEmpty(cypherQuery)) {
```
Comment (felixcheung) [inline neo4j/src/main/java/org/apache/zeppelin/graph/neo4j/Neo4jCypherInterpreter.java]: check isBlank instead

File: zeppelin-distribution/src/bin_license/LICENSE [hunk 9]
```diff
@@ -217,6 +217,7 @@ The following components are provided under Apache License.
     (Apache 2.0) frontend-maven-plugin 1.3 (com.github.eirslett:frontend-maven-plugin:1.3 - https://github.com/eirslett/frontend-maven-plugin/blob/frontend-plugins-1.3/LICENSE
     (Apache 2.0) frontend-plugin-core 1.3 (com.github.eirslett:frontend-plugin-core) - https://github.com/eirslett/frontend-maven-plugin/blob/frontend-plugins-1.3/LICENSE
     (Apache 2.0) mongo-java-driver 3.4.1 (org.mongodb:mongo-java-driver:3.4.1) - https://github.com/mongodb/mongo-java-driver/blob/master/LICENSE.txt
+    (Apache 2.0) Neo4j Java Driver (https://github.com/neo4j/neo4j-java-driver) - https://github.com/neo4j/neo4j-java-driver/blob/1.1/LICENSE.txt
```
Comment (felixcheung) [inline zeppelin-distribution/src/bin_license/LICENSE]: I think we need to include `neo4j-harness` here

Comment (felixcheung) [inline zeppelin-distribution/src/bin_license/LICENSE]: we have version `1.4.0` in the pom file - should we have the same version for license instead of 1.1?

File: neo4j/src/main/java/org/apache/zeppelin/graph/neo4j/Neo4jConnectionManager.java (hunk 3 above)
Comment (conker84) [inline neo4j/src/main/java/org/apache/zeppelin/graph/neo4j/Neo4jConnectionManager.java]: You are totally right, it is my forgetfulness.
Removed

File: zeppelin-distribution/src/bin_license/LICENSE (hunk 9 above)
Comment (conker84) [inline zeppelin-distribution/src/bin_license/LICENSE]: Hi @felixcheung ,
in the #1582 Moon said that _"...org.neo4j.test:neo4j-harness is 'test' scope and will not bundled into the binary package, we don't need include it in zeppelin-distribution/src/bin_license/LICENSE file..."_

Are you sure?

File: docs/interpreter/neo4j.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+limitations under the License.
+-->
+{% include JB/setup %}
+
+# Neo4j Interpreter for Apache Zeppelin
+
+<div id="toc"></div>
+
+## Overview
+[Neo4j](https://neo4j.com/product/) is a native graph database, designed to store and process graphs from bottom to top. On the other hand, non-native solutions only add a shallow graph processing layer to an RDBMS or other NoSQL data stores, resulting in sub-optimal performance.
```
Comment (conker84) [inline docs/interpreter/neo4j.md]: Done!

**Comment by jexp** [comment]
@conker84 ping :)

**Comment by conker84** [comment]
@felixcheung  let me know if i have to do something else!

**Comment by conker84** [comment]
Hi guys, any news/plan about this?

**Comment by felixcheung** [comment]
sorry for dropping this. will review shortly.

**Comment by jexp** [comment]
@felixcheung thank you. We've also received requests from others waiting for this PR :)

File: neo4j/src/main/resources/interpreter-setting.json [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+      },
+      "neo4j.max.concurrency": {
+        "envName": null,
+        "propertyName": "neo4j.max.concurrency",
+        "defaultValue": "50",
+        "description": "Max concurrency call from Zeppelin to Neo4j server."
+      }
+    },
+    "editor": {
+      "language": "cypher",
```
Comment (felixcheung) [inline neo4j/src/main/resources/interpreter-setting.json]: note: there is no syntax highlighting support for cypher, so this won't do anything...

**Comment by conker84** [comment]
Sorry for the late response, i lost it :(
@jexp what do you think about this
_(optional) shall we move neo4j-java-driver to 1.4.3, and neo4j 3.2.3?_
@felixcheung i'm updating the images :)

**Comment by jexp** [comment]
Yes please update.

**Comment by conker84** [comment]
@felixcheung travis is [failing](https://travis-ci.org/conker84/zeppelin), have i to rebase?

**Comment by felixcheung** [comment]
Let's try rebasing

**Comment by felixcheung** [comment]
I've reviewed the changes

**Comment by conker84** [comment]
@felixcheung the job [#10.2](https://travis-ci.org/conker84/zeppelin/jobs/282182858) fails...
Can you help me?

**Comment by jexp** [comment]
@felixcheung ping :)

**Comment by conker84** [comment]
Hi @felixcheung any news?
Thanks
Andrea

**Comment by felixcheung** [comment]
could you rebase to pick up the change, I think the test failure is fixed by PR # 2609

**Comment by felixcheung** [comment]
since this is a known failure that is already fixed, let's not restart the process.
merging if no more comment

**Comment by jexp** [comment]
yay!! thanks so much.

**Comment by conker84** [comment]
Thanks!

**Comment by felixcheung** [comment]
merged, thanks!

**Comment by jexp** [comment]
@felixcheung just curious when is the next Zeppelin release planned ? :)
Can't wait to see this out in the wild.

**Comment by felixcheung** [comment]
Not sure :)

You can ask on dev@zeppelin.apache.org<mailto:dev@zeppelin.apache.org>
